### PR TITLE
[search-parts] Sanitized HTML in templates + CSS prefixing

### PR DESCRIPTION
**Pull request by FranckyC**
- Added sanitization with DOMPurify for every dangerouslySetInnerHTML
- Added CSS prefixing with Web Part instance ID for custom CSS in templates. Support regular CSS rules + Media queries.

### CODE REVIEW

**Comment by FranckyC** [comment]
@wobba If you could review this one to make sure we are aligned.

File: docs/search-parts/templating.md [hunk 1]
```diff
@@ -145,6 +145,14 @@ If provided layouts don't meet your requirements, you can modifiy them or start
 
 ![Edit Template](../images/edit_template.png)
 
+### Styling
+
+You can write your own CSS styles inside templates. However, all CSS rules (including `@media` rules) will be prefixed automatically by an unique ID according to the follwoing pattern (**pnp-modern-search-template_\<Web Part instance ID\>**) to make sure styles are isolated fronm other Web Parts on the page.
```
Comment (wobba) [inline docs/search-parts/templating.md:150]: Spelling error "fronm"

File: search-parts/src/components/BaseWebComponent.ts [hunk 2]
```diff
@@ -34,11 +34,17 @@ export abstract class BaseWebComponent extends HTMLElement {
...
-                    props[camelCase(attr)] = (value === 'true');
-                } else {
-                    props[camelCase(attr)] = value;
+                // Resolve 'data-*' attribute name
+                const matches = attr.match(/data-(.+)/);
+                if (matches && matches.length === 2) {
+                    attr = matches[1];
+
+                    // Booleans
+                    if (/^(true|false)$/.test(value)) {
```
Comment (wobba) [inline search-parts/src/components/BaseWebComponent.ts:43]: Should this regex be case insensitive?

**Comment by wobba** [comment]
Looks good, but I will do some more tests. Will the ''data" prefix be breaking? If so we should consider how to version it, or if we can auto handle upgrading.

File: search-parts/src/webparts/searchResults/components/Layouts/SearchResultsTemplate.tsx [hunk 3]
```diff
@@ -41,7 +62,55 @@ export default class SearchResultsTemplate extends React.Component<ISearchResult
         let templateContent = props.templateContent;
 
         // Process the Handlebars template
-        const template = await this.props.templateService.processTemplate(props.templateContext, templateContent);
+        let template = await this.props.templateService.processTemplate(props.templateContext, templateContent);
+
+        if (template) {
+
+            // Sanitize the template HTML
+            template = template ? this._domPurify.sanitize(`${template}`) : template;
```
Comment (wobba) [inline search-parts/src/webparts/searchResults/components/Layouts/SearchResultsTemplate.tsx:70]: Do you need inline if when you have if check on line above?

Comment (FranckyC) [reply search-parts/src/webparts/searchResults/components/Layouts/SearchResultsTemplate.tsx:70]: Probably not ;)

File: search-parts/src/components/BaseWebComponent.ts (hunk 2 above)
Comment (FranckyC) [reply search-parts/src/components/BaseWebComponent.ts:43]: Right this could be /i

**Comment by FranckyC** [comment]
If people used custom web components, yes it will break. I can remove the check of `data-` attributes if too critical.

**Comment by wobba** [comment]
For custom components it should be ok. What if someone used the oob components manually in a template? Maybe we need to mark some functionality as preview to iron out issues when weintroduce it, and still keep versioning simple :)

**Pull request merged**